### PR TITLE
Keep Kestrel's connection PipeReader in a consistent state

### PR DESCRIPTION
**Pull request by halter73**
- When the request body PipeReader.ReadAsync throws, the connection-level
pipe should be advanced, so subsequent attempts to read from the
connection-level pipe don't fail unnecessarily

Addresses #14727

### CODE REVIEW

**Review by jkotalik — COMMENTED** [review]
Pending tests, this looks good.

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs [hunk 1]
```diff
@@ -70,44 +70,46 @@ public override async ValueTask<ReadResult> ReadAsyncInternal(CancellationToken
...
-
-                CheckCompletedReadResult(_readResult);
-
-                if (readableBufferLength > 0)
+                // Ignore the canceled readResult if it wasn't canceled by the user.
+                if ((!_readResult.IsCanceled && _readResult.Buffer.Length > 0) ||
+                    (_readResult.IsCanceled && Interlocked.Exchange(ref _userCanceled, 0) == 1))
                 {
                     CreateReadResultFromConnectionReadResult();
+                    StopTimingRead(_readResult.Buffer.Length);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs]: Note to self: don't recompute the buffer length.

Comment (jkotalik) [reply src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs]: Where are you computing the buffer length here?

**Comment by anurse** [comment]
Needs @Pilchie approval to get it in, and we need to get it in today for preview3 (without QB approval).

File: src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs (hunk 1 above)
Comment (halter73) [reply src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs]: We were before too. It was just stored into a local before so it wasn't recomputed.

**Comment by halter73** [comment]
@aspnet-hello triage https://dev.azure.com/dnceng/public/_build/results?buildId=412158

**Comment by aspnet-hello** [comment]
This comment was made automatically. If there is a problem contact aspnetcore-build@microsoft.com.

I've triaged the above build. I've created/commented on the following issue(s)
https://github.com/aspnet/AspNetCore-Internal/issues/3154

**Comment by halter73** [comment]
@jkotalik this is ready.

**Review by jkotalik — APPROVED** [review]

**Pull request merged**